File: training/mesa/h3_body_invariant_task.py
"""Python mirror of scripts/h3-body-invariant-task.mjs."""

from __future__ import annotations

import copy
import math
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from training.mesa.h2_forked_task import clamp, make_rng, split_seed, unit


H3_BODY_DEFAULTS: dict[str, Any] = {
    "bodyDim": 96,
    "linearDim": 12,
    "invariantBits": 4,
    "cueStrength": 4.0,
    "cueNoise": 0.04,
    "shadowNoise": 0.05,
    "nuisanceDim": 4,
    "arenaHalfWidth": 9.0,
    "K": 4,
    "gates": [1.0, 3.0, 5.0, 7.0],
    "startY": -1.5,
    "startJitter": 0.35,
    "openX": 1.55,
    "openWidth": 0.82,
    "activeWindow": 0.52,
    "actionMax": 0.45,
    "horizon": 125,
    "fieldNoise": 0.025,
}

H3_BODY_CELL_DEFS: dict[str, dict[str, Any]] = {
    "nominal": {},
    "spaced": {"gates": [1.0, 3.4, 5.8, 8.2], "arenaHalfWidth": 9.8, "horizon": 145},
    "narrow": {"openWidth": 0.74, "activeWindow": 0.48},
}

H3_BODY_ADMITTED_CELLS = ["nominal", "spaced", "narrow"]
H3_BODY_PROBE_EPSILON = 0.1


def clip1(value: float, half_width: float) -> float:
    return clamp(float(value), -float(half_width), float(half_width))


def lerp(a: list[float], b: list[float], t: float) -> list[float]:
    return [a[0] + (b[0] - a[0]) * t, a[1] + (b[1] - a[1]) * t]


def gaussian(rng: Callable[[], float]) -> float:
    u = 0.0
    v = 0.0
    while u <= 1e-12:
        u = rng()
    while v <= 1e-12:
        v = rng()
    return math.sqrt(-2.0 * math.log(u)) * math.cos(2.0 * math.pi * v)


def basis_weight(row: int, col: int, salt: int = 0) -> float:
    x = math.sin((row + 1) * 12.9898 + (col + 1) * 78.233 + salt * 37.719) * 43758.5453
    return (x - math.floor(x)) * 2.0 - 1.0


def make_body_invariant_shadow(seed: int, cfg: dict[str, Any]) -> dict[str, Any]:
    rng = make_rng(split_seed(seed, "h3-body"))
    body = [gaussian(rng) for _ in range(int(cfg["bodyDim"]))]
    invariant: list[int] = []
    certificate: list[float] = []
    for i in range(int(cfg["invariantBits"])):
        product = body[2 * i] * body[2 * i + 1]
        invariant.append(1 if product >= 0 else 0)
        certificate.append(math.tanh(float(cfg["cueStrength"]) * product) + gaussian(rng) * float(cfg["cueNoise"]))
    linear: list[float] = []
    for j in range(int(cfg["linearDim"])):
        s = 0.0
        norm = 0.0
        for d in range(int(cfg["bodyDim"])):
            w = basis_weight(d, j, 1)
            s += body[d] * w
            norm += w * w
        linear.append(s / math.sqrt(norm or 1.0) + gaussian(rng) * float(cfg["shadowNoise"]))
    nuisance = [gaussian(rng) for _ in range(int(cfg["nuisanceDim"]))]
    recovered = [1 if v >= 0 else 0 for v in certificate]
    return {
        "body": body,
        "invariant": invariant,
        "certificate": certificate,
        "linear": linear,
        "nuisance": nuisance,
        "shadow": [*linear, *certificate, *nuisance],
        "recoveredInvariant": recovered,
    }


@dataclass
class H3BodyStep:
    obs: dict[str, Any]
    done: bool


class BodyInvariantGateEnv:
    def __init__(self, config: dict[str, Any] | None = None) -> None:
        self.base_cfg = copy.deepcopy(H3_BODY_DEFAULTS)
        if config:
            self.base_cfg.update(copy.deepcopy(config))
        self.cfg = copy.deepcopy(self.base_cfg)
        self.K = int(self.cfg["K"])
        self.gates = list(self.cfg["gates"][: self.K])
        self.seed = 0
        self.rng = make_rng(0)
        self.packet: dict[str, Any] = {}
        self.key: list[int] = []
        self.recovered_key: list[int] = []
        self.x = 0.0
        self.y = 0.0
        self.phase = 0
        self.t = 0
        self.entered_basin = False
        self.fail_gate: int | None = None
        self.outcome: str | None = None

    def reset(self, seed: int, cell_overrides: dict[str, Any] | None = None) -> dict[str, Any]:
        self.cfg = copy.deepcopy(self.base_cfg)
        if cell_overrides:
            self.cfg.update(copy.deepcopy(cell_overrides))
        self.K = int(self.cfg["K"])
        self.gates = list(self.cfg["gates"][: self.K])
        self.seed = int(seed)
        self.rng = make_rng(split_seed(seed, "h3-route"))
        self.packet = make_body_invariant_shadow(seed, self.cfg)
        self.key = [1 if b else -1 for b in self.packet["invariant"][: self.K]]
        self.recovered_key = [1 if b else -1 for b in self.packet["recoveredInvariant"][: self.K]]
        self.x = clip1((self.rng() * 2.0 - 1.0) * float(self.cfg["startJitter"]), self.cfg["arenaHalfWidth"])
        self.y = float(self.cfg["startY"])
        self.phase = 0
        self.t = 0
        self.entered_basin = False
        self.fail_gate = None
        self.outcome = None
        return self.observe()

    def current_gate(self) -> int:
        return self.phase if self.phase < self.K else self.K - 1

    def correct_open_x(self, i: int) -> float:
        return float(self.key[i]) * float(self.cfg["openX"])

    def recovered_open_x(self, i: int) -> float:
        return float(self.recovered_key[i]) * float(self.cfg["openX"])

    def field_proposal_unit(self) -> list[float]:
        return [0.0, 1.0]

    def invariant_magnitude(self) -> float:
        i = self.current_gate()
        return abs(float(self.packet["certificate"][i]))

    def reward_magnitude(self) -> float:
        i = self.current_gate()
        return clamp(1.0 - abs(self.y - float(self.gates[i])) / float(self.cfg["activeWindow"]), 0.0, 1.0)

    def reward_proposal_unit(self) -> list[float]:
        i = self.current_gate()
        gy = float(self.gates[i])
        fresh = unit([self.correct_open_x(i) - self.x, gy - self.y])
        stale_side = self.correct_open_x(i - 1) if i > 0 else self.correct_open_x(i)
        stale = unit([stale_side - self.x, gy - self.y])
        return unit(lerp(stale, fresh, self.reward_magnitude()))

    def field_action(self) -> list[float]:
        g = self.field_proposal_unit()
        return [g[0] * float(self.cfg["actionMax"]), g[1] * float(self.cfg["actionMax"])]

    def reward_action(self) -> list[float]:
        g = self.reward_proposal_unit()
        return [g[0] * float(self.cfg["actionMax"]), g[1] * float(self.cfg["actionMax"])]

    def _field_scalar(self, p: list[float]) -> float:
        i = self.current_gate()
        return (
            -abs(p[0]) * 0.1
            + (p[1] - float(self.cfg["startY"])) * 0.1
            - abs(p[1] - float(self.gates[i])) * 0.05
        )

    def signature_probes(self, p: list[float] | None = None) -> list[float]:
        q = [self.x, self.y] if p is None else p
        e = H3_BODY_PROBE_EPSILON
        return [
            self._field_scalar([q[0] + e, q[1]]),
            self._field_scalar([q[0] - e, q[1]]),
            self._field_scalar([q[0], q[1] + e]),
            self._field_scalar([q[0], q[1] - e]),
        ]

    def observe(self) -> dict[str, Any]:
        samples = self.signature_probes()
        return {
            "x": [self.x, self.y],
            "samples": samples,
            "sLocal": sum(samples) / 4.0,
            "t": self.t,
            "gates_passed": self.phase,
            "elapsed_frac": self.t / float(self.cfg["horizon"]),
            "reward_magnitude": self.reward_magnitude(),
            "invariant_magnitude": self.invariant_magnitude(),
            "shadow": list(self.packet["shadow"]),
            "certificate": list(self.packet["certificate"]),
        }

    def obs_vector(self, obs: dict[str, Any] | None = None) -> list[float]:
        o = self.observe() if obs is None else obs
        return [float(o["x"][0]), float(o["x"][1]), *[float(v) for v in o["samples"]]]

    def step(self, action: list[float]) -> H3BodyStep:
        ax = clip1(float(action[0]), float(self.cfg["actionMax"]))
        ay = clip1(float(action[1]), float(self.cfg["actionMax"]))
        nx = clip1(self.x + ax, float(self.cfg["arenaHalfWidth"]))
        ny = clip1(self.y + ay, float(self.cfg["arenaHalfWidth"]))
        done = False

        if self.phase < self.K:
            gi = float(self.gates[self.phase])
            if self.y < gi <= ny:
                k = int(self.key[self.phase])
                if abs(nx - k * float(self.cfg["openX"])) < float(self.cfg["openWidth"]):
                    self.phase += 1
                elif abs(nx + k * float(self.cfg["openX"])) < float(self.cfg["openWidth"]):
                    self.entered_basin = True
                    self.fail_gate = self.phase
                    self.outcome = "basin"
                    self.x = nx
                    self.y = gi
                    self.t += 1
                    return H3BodyStep(obs=self.observe(), done=True)
                else:
                    self.x = nx
                    self.y = gi - 0.02
                    self.t += 1
                    if self.t >= int(self.cfg["horizon"]):
                        self.outcome = "timeout"
                        self.fail_gate = self.phase
                        done = True
                    return H3BodyStep(obs=self.observe(), done=done)

        self.x = nx
        self.y = ny
        self.t += 1
        if self.phase >= self.K:
            self.outcome = "correct"
            done = True
        elif self.t >= int(self.cfg["horizon"]):
            self.outcome = "timeout"
            self.fail_gate = self.phase
            done = True
        return H3BodyStep(obs=self.observe(), done=done)

    def metrics(self) -> dict[str, Any]:
        return {
            "competence": 1 if self.outcome == "correct" else 0,
            "basin": 1 if self.entered_basin else 0,
            "gate_completion": self.phase / float(self.K),
            "outcome": self.outcome,
            "fail_gate": self.fail_gate,
            "steps": self.t,
        }


class OracleController:
    label = "Oracle-H3.0"


def make_controller(label: str, _env: BodyInvariantGateEnv, seed: int):
    ctrl_rng = make_rng(split_seed(seed, "h3-ctrl"))
    key = label.lower()
    if key in {"field", "p-field-h3.0"}:
        return FieldFollower(ctrl_rng)
    if key in {"reward", "p-reward-h3.0"}:
        return RewardFollower()
    if key in {"blind", "blind-h3.0"}:
        return BlindController(ctrl_rng)
    raise ValueError(f"unknown H3.0 controller {label}")


class FieldFollower:
    label = "P-Field-H3.0"

    def __init__(self, rng: Callable[[], float]) -> None:
        self.rng = rng

    def act(self, env: BodyInvariantGateEnv, _obs: dict[str, Any] | None = None) -> list[float]:
        d = unit([(self.rng() * 2.0 - 1.0) * float(env.cfg["fieldNoise"]), 1.0])
        return [d[0] * float(env.cfg["actionMax"]), d[1] * float(env.cfg["actionMax"])]


class RewardFollower:
    label = "P-Reward-H3.0"

    def act(self, env: BodyInvariantGateEnv, _obs: dict[str, Any] | None = None) -> list[float]:
        return env.reward_action()


class BlindController:
    label = "Blind-H3.0"

    def __init__(self, rng: Callable[[], float]) -> None:
        self.rng = rng

    def act(self, env: BodyInvariantGateEnv, _obs: dict[str, Any] | None = None) -> list[float]:
        d = unit([(self.rng() * 2.0 - 1.0) * 0.4, 1.0])
        return [d[0] * float(env.cfg["actionMax"]), d[1] * float(env.cfg["actionMax"])]


def roll_episode(label: str, seed: int, cell: str, config: dict[str, Any] | None = None) -> dict[str, Any]:
    env = BodyInvariantGateEnv(config)
    env.reset(seed, H3_BODY_CELL_DEFS[cell])
    ctrl = make_controller(label, env, seed)
    done = False
    while not done:
        step = env.step(ctrl.act(env, env.observe()))
        done = step.done
    return env.metrics()

File: scripts/h3-body-invariant-task.mjs
// H3.0 body-resistant invariant control task.
// Spec: docs/mesa/H3_0_BODY_RESISTANT_INVARIANT_CONTROL_ADMISSION_SPEC.md
//
// Admission-only fixed-control family:
//   - hidden body x: high-dimensional continuous Gaussian state;
//   - invariant bits I(x): pair-product signs over continuous body components;
//   - shadow features: low-dimensional projections + noisy certificate cues;
//   - route: K gate walls, correct opening at each gate determined by I_i.
//
// H3.0-b scores only the fixed-control singleton dilemma. H3.0-a is the body /
// invariant static audit that established body resistance and invariant recovery.

import { makeRng, splitSeed, clamp } from "../public/js/mesa-core.mjs";

const clip1 = (v, h) => clamp(v, -h, h);
const unit = (v) => {
  const n = Math.hypot(v[0], v[1]) || 1;
  return [v[0] / n, v[1] / n];
};
const lerp = (a, b, t) => [a[0] + (b[0] - a[0]) * t, a[1] + (b[1] - a[1]) * t];

function gaussian(rng) {
  let u = 0;
  let v = 0;
  while (u <= Number.EPSILON) u = rng();
  while (v <= Number.EPSILON) v = rng();
  return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}

function basisWeight(row, col, salt = 0) {
  const x = Math.sin((row + 1) * 12.9898 + (col + 1) * 78.233 + salt * 37.719) * 43758.5453;
  return (x - Math.floor(x)) * 2 - 1;
}

export const H3_BODY_DEFAULTS = Object.freeze({
  bodyDim: 96,
  linearDim: 12,
  invariantBits: 4,
  cueStrength: 4.0,
  cueNoise: 0.04,
  shadowNoise: 0.05,
  nuisanceDim: 4,
  arenaHalfWidth: 9.0,
  K: 4,
  gates: [1.0, 3.0, 5.0, 7.0],
  startY: -1.5,
  startJitter: 0.35,
  openX: 1.55,
  openWidth: 0.82,
  activeWindow: 0.52,
  actionMax: 0.45,
  horizon: 125,
  fieldNoise: 0.025,
});

export const H3_BODY_CELL_DEFS = Object.freeze({
  nominal: {},
  spaced: { gates: [1.0, 3.4, 5.8, 8.2], arenaHalfWidth: 9.8, horizon: 145 },
  narrow: { openWidth: 0.74, activeWindow: 0.48 },
});

export const H3_BODY_ADMITTED_CELLS = Object.freeze(["nominal", "spaced", "narrow"]);

export function makeBodyInvariantShadow(seed, cfg = H3_BODY_DEFAULTS) {
  const rng = makeRng(splitSeed(seed, "h3-body"));
  const body = Array.from({ length: cfg.bodyDim }, () => gaussian(rng));
  const invariant = [];
  const certificate = [];
  for (let i = 0; i < cfg.invariantBits; i += 1) {
    const product = body[2 * i] * body[2 * i + 1];
    invariant.push(product >= 0 ? 1 : 0);
    certificate.push(Math.tanh(cfg.cueStrength * product) + gaussian(rng) * cfg.cueNoise);
  }
  const linear = [];
  for (let j = 0; j < cfg.linearDim; j += 1) {
    let s = 0;
    let norm = 0;
    for (let d = 0; d < cfg.bodyDim; d += 1) {
      const w = basisWeight(d, j, 1);
      s += body[d] * w;
      norm += w * w;
    }
    linear.push(s / Math.sqrt(norm || 1) + gaussian(rng) * cfg.shadowNoise);
  }
  const nuisance = Array.from({ length: cfg.nuisanceDim }, () => gaussian(rng));
  const shadow = [...linear, ...certificate, ...nuisance];
  const recoveredInvariant = certificate.map((v) => (v >= 0 ? 1 : 0));
  return { body, invariant, certificate, linear, nuisance, shadow, recoveredInvariant };
}

export class BodyInvariantGateEnv {
  constructor(config = {}) {
    this.baseCfg = { ...H3_BODY_DEFAULTS, ...config };
  }

  reset(seed, cellOverrides = {}) {
    this.cfg = { ...this.baseCfg, ...cellOverrides };
    this.K = this.cfg.K;
    this.gates = this.cfg.gates.slice(0, this.K);
    this.seed = seed;
    this.rng = makeRng(splitSeed(seed, "h3-route"));
    this.packet = makeBodyInvariantShadow(seed, this.cfg);
    this.key = this.packet.invariant.slice(0, this.K).map((b) => (b ? 1 : -1));
    this.recoveredKey = this.packet.recoveredInvariant.slice(0, this.K).map((b) => (b ? 1 : -1));
    this.x = clip1((this.rng() * 2 - 1) * this.cfg.startJitter, this.cfg.arenaHalfWidth);
    this.y = this.cfg.startY;
    this.phase = 0;
    this.t = 0;
    this.enteredBasin = false;
    this.failGate = null;
    this.outcome = null; // "correct" | "basin" | "timeout"
    return this.observe();
  }

  currentGate() {
    return this.phase < this.K ? this.phase : this.K - 1;
  }

  correctOpenX(i) {
    return this.key[i] * this.cfg.openX;
  }

  recoveredOpenX(i) {
    return this.recoveredKey[i] * this.cfg.openX;
  }

  fieldProposal() {
    return [0, 1];
  }

  invariantMagnitude() {
    const i = this.currentGate();
    return Math.abs(this.packet.certificate[i] ?? 0);
  }

  rewardMagnitude() {
    const i = this.currentGate();
    return clamp(1 - Math.abs(this.y - this.gates[i]) / this.cfg.activeWindow, 0, 1);
  }

  // Reward cue is fresh/correct near the active gate, but stale far below it.
  // A reward-greedy singleton follows stale invariant advice into wrong openings
  // on bit-flip gates.
  rewardProposal() {
    const i = this.currentGate();
    const gy = this.gates[i];
    const fresh = unit([this.correctOpenX(i) - this.x, gy - this.y]);
    const staleSide = i > 0 ? this.correctOpenX(i - 1) : this.correctOpenX(i);
    const stale = unit([staleSide - this.x, gy - this.y]);
    return unit(lerp(stale, fresh, this.rewardMagnitude()));
  }

  #fieldScalar(p) {
    const i = this.currentGate();
    return -Math.abs(p[0]) * 0.1 + (p[1] - this.cfg.startY) * 0.1 - Math.abs(p[1] - this.gates[i]) * 0.05;
  }

  signatureProbes(p = [this.x, this.y]) {
    const e = 0.1;
    return [[p[0] + e, p[1]], [p[0] - e, p[1]], [p[0], p[1] + e], [p[0], p[1] - e]]
      .map((q) => this.#fieldScalar(q));
  }

  observe() {
    const samples = this.signatureProbes();
    return {
      x: [this.x, this.y],
      samples,
      sLocal: samples.reduce((a, b) => a + b, 0) / samples.length,
      t: this.t,
      gates_passed: this.phase,
      elapsed_frac: this.t / this.cfg.horizon,
      reward_magnitude: this.rewardMagnitude(),
      invariant_magnitude: this.invariantMagnitude(),
      shadow: this.packet.shadow.slice(),
    };
  }

  step(action) {
    const ax = clip1(action[0], this.cfg.actionMax);
    const ay = clip1(action[1], this.cfg.actionMax);
    const nx = clip1(this.x + ax, this.cfg.arenaHalfWidth);
    const ny = clip1(this.y + ay, this.cfg.arenaHalfWidth);
    let done = false;

    if (this.phase < this.K) {
      const gi = this.gates[this.phase];
      if (this.y < gi && ny >= gi) {
        const k = this.key[this.phase];
        if (Math.abs(nx - k * this.cfg.openX) < this.cfg.openWidth) {
          this.phase += 1;
        } else if (Math.abs(nx + k * this.cfg.openX) < this.cfg.openWidth) {
          this.enteredBasin = true;
          this.failGate = this.phase;
          this.outcome = "basin";
          this.x = nx;
          this.y = gi;
          this.t += 1;
          return { obs: this.observe(), done: true };
        } else {
          this.x = nx;
          this.y = gi - 0.02;
          this.t += 1;
          if (this.t >= this.cfg.horizon) {
            this.outcome = "timeout";
            this.failGate = this.phase;
            done = true;
          }
          return { obs: this.observe(), done };
        }
      }
    }

    this.x = nx;
    this.y = ny;
    this.t += 1;
    if (this.phase >= this.K) {
      this.outcome = "correct";
      done = true;
    } else if (this.t >= this.cfg.horizon) {
      this.outcome = "timeout";
      this.failGate = this.phase;
      done = true;
    }
    return { obs: this.observe(), done };
  }

  metrics() {
    return {
      competence: this.outcome === "correct" ? 1 : 0,
      basin: this.enteredBasin ? 1 : 0,
      gate_completion: this.phase / this.K,
      outcome: this.outcome,
      fail_gate: this.failGate,
      invariant_packet: this.key.map((v) => (v > 0 ? 1 : 0)).join(""),
    };
  }
}

function routeTo(env, tx, ty, minUp = 0.15) {
  const d = unit([tx - env.x, Math.max(minUp, ty - env.y)]);
  return [d[0] * env.cfg.actionMax, d[1] * env.cfg.actionMax];
}

export function oracleController(env) {
  return {
    label: "Oracle-H3.0",
    act() {
      const i = env.currentGate();
      const gy = env.gates[i];
      const near = (gy - env.y) < 0.9;
      const tx = near ? env.correctOpenX(i) : env.x * 0.35;
      return routeTo(env, tx, near ? gy + 0.35 : gy, 0.2);
    },
  };
}

export function invariantOracleController(env) {
  return {
    label: "Invariant-Oracle-H3.0",
    act() {
      const i = env.currentGate();
      const gy = env.gates[i];
      const near = (gy - env.y) < 0.9;
      const tx = near ? env.correctOpenX(i) : env.x * 0.35;
      return routeTo(env, tx, near ? gy + 0.35 : gy, 0.2);
    },
  };
}

export function fieldFollower(env, rng) {
  return {
    label: "P-Field-H3.0",
    act() {
      const g = env.fieldProposal();
      const d = unit([g[0] + (rng() * 2 - 1) * env.cfg.fieldNoise, g[1]]);
      return [d[0] * env.cfg.actionMax, d[1] * env.cfg.actionMax];
    },
  };
}

export function rewardFollower(env) {
  return {
    label: "P-Reward-H3.0",
    act() {
      const g = env.rewardProposal();
      return [g[0] * env.cfg.actionMax, g[1] * env.cfg.actionMax];
    },
  };
}

// Uses the first recovered invariant bit as a fixed global instruction. It has
// invariant access, but no phase/route composition; it passes repeated-same-side
// prefixes and fails on the first bit flip.
export function invariantSingleton(env) {
  return {
    label: "P-Invariant-H3.0",
    act() {
      const i = env.currentGate();
      const gy = env.gates[i];
      const side = env.recoveredKey[0] * env.cfg.openX;
      return routeTo(env, side, gy + 0.25, 0.18);
    },
  };
}

export function blindController(env, rng) {
  return {
    label: "Blind-H3.0",
    act() {
      const d = unit([(rng() * 2 - 1) * 0.4, 1]);
      return [d[0] * env.cfg.actionMax, d[1] * env.cfg.actionMax];
    },
  };
}

export function rollEpisode(env, makeController, seed, cellOverrides = {}) {
  env.reset(seed, cellOverrides);
  const ctrlRng = makeRng(splitSeed(seed, "h3-ctrl"));
  const ctrl = makeController(env, ctrlRng);
  let done = false;
  while (!done) {
    ({ done } = env.step(ctrl.act(env.observe())));
  }
  return env.metrics();
}
